add pre assessment revisions

RoordinkFrank opened the pull request:
After a pre assesment revision has been submitted the secretary view looks like this:
<img width="1209" height="712" alt="image" src="https://github.com/user-attachments/assets/29cf3982-62c7-4ea4-9ee3-d2fd67d1c950" />
change comparison looks like this:
<img width="1467" height="951" alt="image" src="https://github.com/user-attachments/assets/81e978d6-e4f3-42dc-ab4f-31b6c2f3013a" />
It does look in order as far as I can see.

backup comments before git reset
<img width="767" height="320" alt="image" src="https://github.com/user-attachments/assets/68df50bb-05b0-4092-a730-92b998ce2127" />
<img width="795" height="222" alt="image" src="https://github.com/user-attachments/assets/9a807b9f-deee-40d4-8c1b-5c408b12e20b" />

RoordinkFrank committed:
Revision is now possible

RoordinkFrank committed:
comment change

RoordinkFrank commented on proposals/forms.py:
```diff
@@ -500,12 +500,11 @@ def __init__(self, *args, **kwargs):
 
     def _get_parent_queryset(self):
-        # Select non-pre-assessments that have been reviewed and rejected and
+        # Select proposals that have been reviewed and rejected and
```
I am still not sure what exactly the difference is between RevisionProposalCopyForm and AmendmentProposalCopyForm . I assume for now that I should make this change for both of them.

miggol replied:
Yes... I suppose preassessments can be amended as well now!

EdoStorm96 reviewed (approved):
Nice catch that you found `is_revisable`! I'd forgotten all about it ...

Looks all good!

RoordinkFrank committed:
Merge branch 'develop' into Feat/pre-assesment_get_revisions

miggol commented on proposals/forms.py:
```diff
@@ -459,9 +459,9 @@ def __init__(self, *args, **kwargs):
+        # Return all proposals, that are not currently in review
         return (
-            Proposal.objects.filter(is_pre_assessment=False)
+            Proposal.objects.filter()
```
Now that we're touching this, could you make this part of the `ProposalQuerySet` manager in `models.py`? What I want is that the manager provides the queryset of all revisable proposals, and this form only filters on the applicant part.

RoordinkFrank replied:
Do you want something like this? And if so just for this method or for others as well?

miggol replied:
What do you mean?

RoordinkFrank replied:
This is about commit https://github.com/DH-IT-Portal-Development/ethics/pull/1001/commits/27c18da1a7dc41f6c2d8abd7cc8340638ef3d661

miggol reviewed (changes requested):
I haven't elaborately tested the copying mechanism, but assuming that works, this should be almost ready for merge. I found some small things that need to be addressed:

- Logic in email templates isn't designed for revised preassessments, for example `assignment_shortroute.txt`
- Emails to secretary will always be for a "New preassessment proposal" rather than reflecting the revised status
- It would be nice to use the Proposal QS manager to return revisable applications and the filter on that basis. But that's ultimately nice-to-have and not important

RoordinkFrank commented:
I have yet to add the tests. Something we discussed verbally. As for your comments I will look into them.

RoordinkFrank commented on reviews/tests.py:
```diff
@@ -151,6 +152,80 @@ def test_start_review(self):
+        print("ReviewPreAssessmentTestCase: Creating pre-assessment proposal")
+        self.setup_pre_assessment()
+
+    def setup_pre_assessment(self):
```
Follows the same setup as above. That said if we want to have more then one example data we may need to load this from an external file (SQL perhaps?)

Goes outside the scope of this issue but something to think about

RoordinkFrank commented on reviews/tests.py:
```diff
@@ -107,15 +108,78 @@ def refresh(self):
+        self.direct_funding: Funding = Funding.objects.get(pk=2)
+        self.setup_pre_assessment()
+
+    def setup_pre_assessment(self):
```
If it turns out this data will be needed in all test classes we may just as well move this to BaseReviewTestCase. Still I am not sure if that's the case.

RoordinkFrank commented on reviews/tests.py:
```diff
@@ -107,15 +108,78 @@ def refresh(self):
+        # No Data yet added
+        # 4. Submit
+        # Space for possible comments. Use a maximum of 1000 words.
+        self.pre_assessment.comments = "Space for possible comments"
```
Not sure if pre_assessment study is needed.

EdoStorm96 reviewed (changes requested):
Well this is a good start and there are some good ideas here. However this can be streamlined a lot. I'd like you to have a look at `proposal_submission_tests.py` and try to replicate the pattern seen there. This can make the code a lot more concise.

One of the main differences there is that the proposals are loaded from fixtures instead of created in the tests. I think you might be able to simply repurpose `BaseProposalTestCase` from the submission test file for the tests that are being run here. From the fixtures we already create a pre_assessment and a pre_approved proposal, so you can the also test pre_approved proposals while you're at it.

To test the different types we don't create multiple functions, but instead create a new subclass where we override self.proposal. I think this is pretty clean. It will be a bit of work to get these tests working with the setup from `BaseProposalTestCase`, but it will be very nice once it does become compatible! (In fact, I was surprised we do not repurpose `BaseProposalTestCase` for these tests already ...)

Let me know if you have any other questions and keep up the good work :)

RoordinkFrank committed:
Mail changes.

RoordinkFrank committed:
correct mail tags

RoordinkFrank committed:
Part of _get_parent_queryset in forms.py moved to the Proposalqueryset.

RoordinkFrank committed:
removed unused imports

RoordinkFrank committed:
black .

RoordinkFrank commented on proposals/models.py:
```diff
@@ -90,6 +90,12 @@ def __str__(self):
     DECISION_MADE = 55
 
+    def all_proposals_not_in_review(self):
+        return self.filter(
```
the .distinct now happens in the middle of the filter instead at the end. I am also not 100% sure if it is needed here.

miggol replied:
I don't think it is needed, but maybe I'm missing something. What makes you think it is needed?

RoordinkFrank replied:
.distinct was in the original query. Except it was at the end and now in the middle. That is a change of logic, that is why I highlighted it. 

As for .distinct itself I am not sure if it is needed either.

edit: removed .distinct() to end of query as was the original logic. No changes to thing I do not fully understand why/if they are needed.

miggol commented on proposals/models.py:
```diff
@@ -90,6 +90,12 @@ def __str__(self):
 class ProposalQuerySet(models.QuerySet):
     DECISION_MADE = 55
 
+    def all_proposals_not_in_review(self):
```
The name of this method obscures its semantics: `copyable_proposals()` would be better. This is where the queryset of copyable proposals lives.

RoordinkFrank replied:
agreed

miggol commented on reviews/templates/mail/concept_supervisor.txt:
```diff
@@ -11,7 +11,9 @@ Je kunt deze concept-aanvraag bekijken via {{ proposal_url }}. Je kunt deze, eve
 
 Je kunt al je aanvragen als eindverantwoordelijke onderzoeker en hun status via {{ my_supervised }} bekijken.
-
+{% if is_pre_assessment %}
+Let op: het gaat hierbij om een pre-assessment.
+{% endif %}
```
This variable isn't provided by `start_supervisor_phase()`. I'm guessing you didn't test this?

Remember to also modify the HTML emails when editing the text emails :)

RoordinkFrank replied:
caught out. I indeed did not test this properly. Also good catch to have found this.

RoordinkFrank committed:
feedback: removed distinct to end of query,
Renamed copyable_proposals, mail not also containts .html. added variable to supervisor mail,

RoordinkFrank committed:
. black

RoordinkFrank committed:
DeeJeeLint

RoordinkFrank committed:
Pre Assessment tests, small msg's in places where the assert does not give enough info. Pre assessment data is partly from fixtures/partly added myself.

RoordinkFrank committed:
comments added

RoordinkFrank committed:
Does not match assert condition message changed. Test auto review pre assessment moved below. Return statement changed to pass according to feedback.

RoordinkFrank committed:
A constant has been introduced to make future changes easier.

RoordinkFrank committed:
small comment

RoordinkFrank committed:
small comment

RoordinkFrank committed:
pre assessment no longer uses fixtures

RoordinkFrank committed:
test_auto_review_short_route moved to test_auto_review.
Removed comment on est_auto_review_session_time
error message changed.

RoordinkFrank committed:
Merge branch 'develop' into Feat/pre-assesment_get_revisions

RoordinkFrank committed:
Merge branch 'Feat/pre-assesment_get_revisions' into feat/pre-assesment_tests

RoordinkFrank committed:
Merge pull request #1016 from DH-IT-Portal-Development/feat/pre-assesment_tests

Pre-Assessment tests

EdoStorm96 commented on reviews/tests.py:
```diff
@@ -76,6 +78,26 @@ def setup_proposal(self):
+            metc=YesNoDoubt.NO,
+        )
+        self.pre_assessment.supervisor = self.supervisor
+        self.pre_assessment.relation = Relation.objects.get(description_en=PHD_STUDENT)
```
I think just getting the relation via the PK is better ... It is more likely to be constant and it is how we roll for the rest of the tests. Then you can remove the test_constants file.

RoordinkFrank replied:
Conflicting feedback here. Though that feedback was in feat/pre-assessment_tests I believe.

Edit found it:

<img width="941" height="672" alt="image" src="https://github.com/user-attachments/assets/9e254b97-ad03-4ae0-8afa-2928b82d7103" />

EdoStorm96 commented on proposals/forms.py:
```diff
@@ -459,19 +459,15 @@ def __init__(self, *args, **kwargs):
+        # Return all proposals, that are not currently in review
         return (
-            Proposal.objects.filter(is_pre_assessment=False)
+            Proposal.objects.copyable_proposals()
```
Right now I don't really like this logic being in a manager, as this only gets used in spot and now I have to visit a different file to see what is happening in that one spot ...

EdoStorm96 replied:
I do however see a use for a `can_be_copied_by(user)` filter, as that gets used in all these querysets? Maybe in more places.

That would replace this part:
```
            .filter(
                Q(
                    applicants=self.user,
                )
                | Q(supervisor=self.user)
            )
```
But you can do this if you want, it's not a must for me.

RoordinkFrank replied:
I can see your point but I do get a bit of conflicting feedback here. I will take a look at can_be_copied_by(user).

<img width="886" height="720" alt="image" src="https://github.com/user-attachments/assets/a41bd00a-7b5e-4279-a1a6-876eed8a85b7" />

EdoStorm96 commented on proposals/models.py:
```diff
@@ -90,6 +90,12 @@ def __str__(self):
     DECISION_MADE = 55
 
+    def copyable_proposals(self):
+        return self.filter(
```
As I mention in another comment, I think having a filter which only gets used in one spot is not so useful. I say, get rid of it.

EdoStorm96 commented on proposals/models.py:
```diff
@@ -681,8 +681,7 @@ def is_revisable(self):
         keep in mind to also check if the user is one of the applicants
         or the supervisor."""
         if (
-            not self.is_pre_assessment
```
Nice

EdoStorm96 commented on reviews/tests.py:
```diff
@@ -76,6 +81,26 @@ def setup_proposal(self):
+            ),
+            institution_id=1,
+        )
+        self.pre_assessment.is_pre_assessment = True
```
Why is this outside of the creation? It can also be done during the creation right?

RoordinkFrank replied:
Done.

EdoStorm96 commented on reviews/tests.py:
```diff
@@ -401,6 +523,44 @@ def test_auto_review_registration_age_min(self):
         self.assertEqual(len(reasons), 2)
 
 
+class PreAssessmentAutoReviewTestCase(AutoReviewTests):
```
Pre assessments don't have a study, so I think this whole thing can go.

RoordinkFrank replied:
This does make Sense.
I added a comment `# pre-assessment does not have a study so PreAssessmentAutoReviewTestCase should and not be added.` If you feel it's unnecessary it can be removed.

EdoStorm96 commented on reviews/tests.py:
```diff
@@ -409,6 +569,9 @@ class ReviewCloseTestCase(
         super().setUp()
+        self.startReview()
+
+    def startReview(self):
```
Naming convention for python class methods is `lower_case_with_underscores`.

RoordinkFrank replied:
Switching between python en typescript is hard:(

EdoStorm96 commented on reviews/tests.py:
```diff
@@ -493,6 +656,7 @@ def test_long_route(self):
         self.assertEqual(
             self.review.stage,
             self.review.Stages.CLOSED,
+            f"Pre-assessment proposal stage is incorrect. We expect {Review.Stages.CLOSED.name}, but instead we got {Review.Stages(self.review.stage).name}",
```
This is not PreAss right?

RoordinkFrank replied:
I have no idea why I wrote Pre-assessment there. It's not even a proposal stage

EdoStorm96 reviewed (changes requested):
There is some small stuff I'd like to see changed. Almost there :)

RoordinkFrank committed:
feedback:
- Removed PreAssessmentAutoReviewTestCase.
- Start_review is now lowercase.
- Error message fixed.
- setup_proposal code has been moved to create objects

RoordinkFrank committed:
feedback:
- Can_be_copied_by(user) added to parent_queryset.

RoordinkFrank committed:
style: returns QuerySet added for clarity.

EdoStorm96 reviewed (approved):
Looks good!

RoordinkFrank committed:
black .
